src/main/java/com/google/sps/placeGuide/repository/impl/DatastorePlaceGuideRepository.java

```diff
@@ -94,6 +127,61 @@ private Entity createPlaceGuideEntity(PlaceGuide placeGuide) {
     return Collections.unmodifiableList(createdPlaceGuides);
   }
 
+  @Override
+  public List<PlaceGuide> getBookmarkedPlaceGuides(String userId) {
+    List<PlaceGuide> bookmarkedPlaceGuides = new ArrayList<>();
+    User user = userRepository.getUser(userId);
+    if (user != null) {
+      Set<Long> bookmarkedIds = user.getBookmarkedPlaceGuidesIds();
+      List<Long> bookmarkedIdsCopy = new ArrayList<>(bookmarkedIds);
+      int bookmarkedIdsCopyIndex = 0;
+      while (bookmarkedIdsCopyIndex < bookmarkedIdsCopy.size()) {
+        long placeGuideId = bookmarkedIdsCopy.get(bookmarkedIdsCopyIndex);
+        Key placeGuideEntityKey = KeyFactory.createKey(ENTITY_KIND, placeGuideId);
+        try {
+          Entity placeGuideEntity = datastore.get(placeGuideEntityKey);
+          bookmarkedPlaceGuides.add(getPlaceGuideFromEntity(placeGuideEntity));
+          bookmarkedIdsCopyIndex++;
+        } catch(EntityNotFoundException err) {
+          System.out.println("PlaceGuide entity does not exist or has already been removed.");
+          bookmarkedIdsCopy.remove(bookmarkedIdsCopyIndex);
+        }
+      }
+      Set<Long> updatedBookmarkedIdsCopy = new HashSet<>(bookmarkedIdsCopy);
+      // Update and save user with updated {@code bookmarkedPlaceGuides}.
+      saveUpdatedUser(user, updatedBookmarkedIdsCopy);
+      return bookmarkedPlaceGuides;
+    } else {
+      throw new IllegalStateException("Cannot get bookmarked place guides for non-existent user!");
+    }
+  }
+
+  @Override
+  public void removeBookmarkedPlaceGuide(long placeGuideId, String userId) {
+    User user = userRepository.getUser(userId);
+    if (user != null) {
+      Set<Long> bookmarkedPlaceGuidesIds = user.getBookmarkedPlaceGuidesIds();
+      Set<Long> bookmarkedPlaceGuidesIdsCopy = new HashSet<>(bookmarkedPlaceGuidesIds);
+      bookmarkedPlaceGuidesIdsCopy.remove(placeGuideId);
+      
+      // Update user with the new set.
+      saveUpdatedUser(user, bookmarkedPlaceGuidesIdsCopy);
+    } else {
+      throw new IllegalStateException("Cannot remove bookmarked place guides from non-existent user!");
+    }
+  }
+
+  private void saveUpdatedUser(User user, Set<Long> updatedBookmarkedPlaceGuidesIds) {
+    User updatedUser =
```

Comment:
Instead of rebuilding the user object, we could implement a function for the User class to add a new placeGuideId to its list/remove one. Since you are using a set, this would be easier and also more efficient.

Reply:
Staying with updating the user with an updated `bookmarkedPlaceGuidesIds` to keep it immutable.